### PR TITLE
[tune] only sync up and sync down checkpoint folder for cloud checkpoint.

### PR DESCRIPTION
**Pull request by xwjiang2010**
By default, `~/ray_results/exp_name/trial_name/checkpoint_name`.

Instead of the whole trial checkpoint (~/ray_results/exp_name/trial_name/) directory.

Stuff like `progress.csv`, `result.json`, `params.pkl`, `params.json`, `events.out` etc are coming from driver process. 

This could also enable us to de-couple `sync up` and `delete` - they don't have to wait for each other to finish. 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by krfricke — COMMENTED** [review]
This looks good. Two questions:

1) Should we also do this in `restore`?
2) Can we run release tests on this to make sure cloud checkpointing behavior still works?

**Comment by xwjiang2010** [comment]
https://buildkite.com/ray-project/periodic-ci/builds/2434

Lint error is unrelated.

File: python/ray/tune/utils/trainable.py [hunk 1]
```diff
@@ -100,6 +115,27 @@ def find_checkpoint_dir(checkpoint_path):
...
+        MyTrainable_a=1_b=2_2022-01-08_20-45-51/`.
+
+        returns `checkpoint_00000`.
+
+        Raises:
+            FileNotFoundError if checkpoint directory is not found.
+        """
+        experiment_dir = os.path.dirname(os.path.dirname(logdir))
+        rel_path = os.path.relpath(checkpoint_path, experiment_dir)
+        return rel_path.split(os.sep)[1]
```
Comment (krfricke) [inline python/ray/tune/utils/trainable.py:137]: This method is unstable and relies on specific input formatting, like trailing slashes. It also doesn't work with checkpoints from different trials or experiments (though the latter is not necessarily required).

Here are some test cases:

```

# With trailing slash --> checkpoint_0 (expected)
print(find_rel_checkpoint_dir("/tmp/exp1/trial0/", "/tmp/exp1/trial0/checkpoint_0"))

# Without trailing slash --> trial0 (should be checkpoint_0)
print(find_rel_checkpoint_dir("/tmp/exp1/trial0", "/tmp/exp1/trial0/checkpoint_0"))

# Different trial with trailing slash --> checkpoint_0 (should be ../trial1/checkpoint_0)
print(find_rel_checkpoint_dir("/tmp/exp1/trial0/", "/tmp/exp1/trial1/checkpoint_0"))

# Different trial without trailing slash --> trial1 (should be ../trial1/checkpoint_0)
print(find_rel_checkpoint_dir("/tmp/exp1/trial0", "/tmp/exp1/trial1/checkpoint_0"))

# Different experiment with trailing slash --> exp1 (should be ../../exp1/trial1/checkpoint_0)
print(find_rel_checkpoint_dir("/tmp/exp2/trial0/", "/tmp/exp1/trial1/checkpoint_0"))

# Different experiment without trailing slash --> trial1 (should be ../../exp1/trial1/checkpoint_0)
print(find_rel_checkpoint_dir("/tmp/exp2/trial0.", "/tmp/exp1/trial1/checkpoint_0"))
```

Adjusting the function like this yields the expected results:

```suggestion
        experiment_dir = os.path.dirname(logdir + os.sep)
        rel_path = os.path.relpath(checkpoint_path, experiment_dir)
        return rel_path
```

Output
```
checkpoint_0
checkpoint_0
../trial1/checkpoint_0
../trial1/checkpoint_0
../../exp1/trial1/checkpoint_0
../../exp1/trial1/checkpoint_0
```

Maybe we should add a unit test for that?

**Review by krfricke — CHANGES_REQUESTED** [review]
Great, let's just fix this helper method

File: python/ray/tune/utils/trainable.py (hunk 1 above)
Comment (xwjiang2010) [reply python/ray/tune/utils/trainable.py]: Added a unit test. Although I think all these cases should just return checkpoint_0.

**Comment by xwjiang2010** [comment]
https://buildkite.com/ray-project/periodic-ci/builds/2445#729c50d2-c98d-4965-93e5-2f32b506c8d8
all passing

File: python/ray/tune/utils/trainable.py (hunk 1 above)
Comment (xwjiang2010) [reply python/ray/tune/utils/trainable.py]: Actually I don't think checkpoint_path and self.logdir may refer to different trials.
What observed before was an artifact of confusing test fixture set up.
So I instead removed this case and updated the test: python/ray/tune/tests/test_api.py::TrainableFunctionApiTest::testDurableTrainableClass/testDurableTrainableFunction.

**Review by krfricke — APPROVED** [review]
Looks good, thanks!

**Pull request merged**